sdk/python/kfp/v2/components/component_factory.py

```diff
@@ -35,7 +35,7 @@ def _python_function_name_to_component_name(name):
 def _get_packages_to_install_command(
         package_list: Optional[List[str]] = None) -> List[str]:
     result = []
-    if package_list is not None:
```

Comment:
I'd argue that it could also be a user explicitly expressing no packages. It's clearer in code than `None` or leaving the default.

In fact I did exactly that which is what prompted the PR.

Reply:
That's not true, see: https://github.com/kubeflow/pipelines/blob/d83933dafd7ae84659c9dff7d170b8655b81df82/sdk/python/kfp/v2/components/component_factory.py#L154

In the code the `None` is replaced with an empty list anyway. So leaving the default `None` will result in the empty list.

Reply:
That's not true, see: https://github.com/kubeflow/pipelines/blob/d83933dafd7ae84659c9dff7d170b8655b81df82/sdk/python/kfp/v2/components/component_factory.py#L154

In the code the `None` is replaced with an empty list anyway. So leaving the default `None` will result in the empty list.

Reply:
I see, maybe an alternative change it to move `packages_to_install = packages_to_install or []` on line 154 to line 158:

https://github.com/kubeflow/pipelines/blob/2f19a26ffd205c7a3384c4c250f5dbb5150a7c4d/sdk/python/kfp/v2/components/component_factory.py#L154-L158

WDYT?

Reply:
Personally I think I prefer the current change and would maybe also remove the `Optional` from `_get_packages_to_install_command`.

Imo `_get_packages_to_install_command` should have enough logic (and does) to identify when there are no packages (empty list). It accepting a 'None` type too just muddies the water in my mind. I'm not sure I really understand the advantage of it taking `None`

Reply:
> It accepting a None type too just muddies the water in my mind. I'm not sure I really understand the advantage of it taking None

Please don't remove the `Optional`. It's not a good idea to use mutable objects as default arguments:
https://google.github.io/styleguide/pyguide.html#212-default-argument-values

Reply:
I personally prefer the alternative change I mentioned above, in which way `packages_to_install == None` indicates system default, whereas `packages_to_install == []` must be from user input. 

That said, I'm fine with the current change as well.

Reply:
@neuromage That's not what I was suggesting at all! I'm fully aware that mutable default args are a very bad idea.

My suggestion was simply to remove the `Optional` type hint on `_get_packages_to_install_command` to make it clear that a list is expected.

Reply:
> It's a private function (as much as such a thing exists in Python) which is only used internally so there's not really a good reason for the Optional in my mind. And it already doesn't have a default value.

`Optional[...]` is just a shorthand for `Union[None, ...]`, it has nothing to do with default value. With the alternative change I commented above, we would get `None`--the default value from the public method--passed all the way through to this point. So `Optional` would make sense.